### PR TITLE
feat(core): impl AI switch

### PR DESCRIPTION
**Pull request by JimmFly**
close PD-1658

https://github.com/user-attachments/assets/2f3d1d26-1879-4d95-b80c-7c0965cefbd0

### CODE REVIEW

**Comment by JimmFly** [comment]
* **#8018** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/8018?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `canary`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @JimmFly and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: packages/frontend/core/src/blocksuite/presets/ai/ai-spec.ts [hunk 1]
```diff
@@ -17,140 +17,164 @@ import {
...
-      ...PageRootBlockSpec.view.widgets,
-      [AFFINE_AI_PANEL_WIDGET]: literal`${unsafeStatic(
-        AFFINE_AI_PANEL_WIDGET
-      )}`,
+const pageWidgets: Record<string, StaticValue> = {
+  ...PageRootBlockSpec.view.widgets,
+};
+
+export const createAIPageRootBlockSpec = (enableAI?: boolean): BlockSpec => {
+  if (enableAI) {
```
Comment (EYHN) [inline packages/frontend/core/src/blocksuite/presets/ai/ai-spec.ts]: can we just not calling this function when enableAI is false？

File: packages/frontend/core/src/blocksuite/presets/ai/entries/code-toolbar/setup-code-toolbar.ts [hunk 2]
```diff
@@ -16,7 +16,10 @@ import type { AskAIButtonOptions } from '../../_common/components/ask-ai-button'
 import { buildAICodeItemGroups } from '../../_common/config';
 import { AIStarIcon } from '../../_common/icons';
 
-export function setupCodeToolbarEntry(codeToolbar: AffineCodeToolbarWidget) {
+export function setupCodeToolbarEntry(
```
Comment (EYHN) [inline packages/frontend/core/src/blocksuite/presets/ai/entries/code-toolbar/setup-code-toolbar.ts]: can we split this function to ai and non-ai ？

File: packages/frontend/core/src/blocksuite/presets/ai/ai-spec.ts (hunk 1 above)
Comment (JimmFly) [reply packages/frontend/core/src/blocksuite/presets/ai/ai-spec.ts]: @L-Sun will follow up on these issues.

File: packages/frontend/core/src/blocksuite/presets/ai/entries/code-toolbar/setup-code-toolbar.ts (hunk 2 above)
Comment (L-Sun) [reply packages/frontend/core/src/blocksuite/presets/ai/entries/code-toolbar/setup-code-toolbar.ts]: Done

File: packages/frontend/core/src/blocksuite/presets/ai/ai-spec.ts (hunk 1 above)
Comment (L-Sun) [reply packages/frontend/core/src/blocksuite/presets/ai/ai-spec.ts]: Done

File: packages/frontend/core/src/components/affine/setting-modal/general-setting/editor/general.tsx [hunk 3]
```diff
@@ -310,13 +311,42 @@ const NewDocDefaultModeSettings = () => {
...
+        cancelText: 'Cancel',
+        onConfirm: () => onAIChange(checked),
+        confirmButtonOptions: {
+          variant: checked ? 'primary' : 'error',
+        },
+      });
+    },
+    [openConfirmModal, onAIChange]
+  );
+
```
Comment (forehalo) [inline packages/frontend/core/src/components/affine/setting-modal/general-setting/editor/general.tsx]: DO NOT BRING CHILDREN STATE TO PARENT ELEMENT

File: packages/frontend/core/src/components/blocksuite/block-suite-editor/lit-adaper.tsx [hunk 4]
```diff
@@ -81,11 +86,15 @@ const usePatchSpecs = (page: Doc, shared: boolean, mode: DocMode) => {
     };
   }, [page.collection]);
 
+  const enableAI = useLiveData(
```
Comment (forehalo) [inline packages/frontend/core/src/components/blocksuite/block-suite-editor/lit-adaper.tsx]: it safe to not use live data here, useless re-render can be avoided if so

File: packages/frontend/core/src/components/blocksuite/block-suite-editor/specs/page.ts [hunk 5]
```diff
@@ -5,15 +5,18 @@ import {
...
 import { type FrameworkProvider } from '@toeverything/infra';
 
-import { CommonBlockSpecs } from './common';
+import { createAISpecs, createDefaultSpecs } from './common';
 import { createPageRootBlockSpec } from './custom/root-block';
 
-export function createPageModeSpecs(framework: FrameworkProvider): BlockSpec[] {
+export function createPageModeSpecs(
+  framework: FrameworkProvider,
+  enableAI?: boolean
```
Comment (forehalo) [inline packages/frontend/core/src/components/blocksuite/block-suite-editor/specs/page.ts]: is it optional?

File: packages/frontend/core/src/components/blocksuite/block-suite-editor/lit-adaper.tsx (hunk 4 above)
Comment (EYHN) [inline packages/frontend/core/src/components/blocksuite/block-suite-editor/lit-adaper.tsx]:
```suggestion
  const [enableAI] = useState(editorSettingService.editorSetting.settings$.value. enableAI)
```

File: packages/frontend/core/src/pages/workspace/detail-page/detail-page.tsx [hunk 6]
```diff
@@ -85,6 +88,10 @@ const DetailPageImpl = memo(function DetailPageImpl() {
   const globalContext = globalContextService.globalContext;
   const doc = docService.doc;
 
+  const enableAI = useLiveData(
```
Comment (EYHN) [inline packages/frontend/core/src/pages/workspace/detail-page/detail-page.tsx]:
```suggestion
  const [enableAI] = useState(editorSettingService.editorSetting.settings$.value. enableAI)
```

File: packages/frontend/core/src/components/affine/setting-modal/general-setting/editor/general.tsx [hunk 7]
```diff
@@ -310,13 +311,42 @@ const NewDocDefaultModeSettings = () => {
 
 export const General = () => {
   const t = useI18n();
+  const { openConfirmModal } = useConfirmModal();
+  const { editorSettingService } = useServices({ EditorSettingService });
+
+  const settings = useLiveData(editorSettingService.editorSetting.settings$);
```
Comment (EYHN) [inline packages/frontend/core/src/components/affine/setting-modal/general-setting/editor/general.tsx:402]:
```suggestion
  const enableAI = useLiveData(editorSettingService.editorSetting.settings$.map(s => s. enableAI));
```

File: packages/frontend/core/src/components/affine/ai-onboarding/index.tsx [hunk 8]
```diff
@@ -28,19 +30,29 @@ const useDismiss = (key: AIOnboardingType) => {
...
+      {!enableAI || dismissGeneral ? null : <AIOnboardingGeneral />}
+      {!enableAI || dismissLocal ? null : <AIOnboardingLocal />}
     </Suspense>
   );
 };
 
 export const PageAIOnboarding = () => {
   const [dismissEdgeless] = useDismiss(AIOnboardingType.EDGELESS);
+  const editorSettingService = useService(EditorSettingService);
+  const enableAI = useLiveData(
```
Comment (EYHN) [inline packages/frontend/core/src/components/affine/ai-onboarding/index.tsx:49]:
```suggestion
  const enableAI = useLiveData(editorSettingService.editorSetting.settings$.map(s => s. enableAI));
```

File: packages/frontend/core/src/components/affine/setting-modal/general-setting/editor/general.tsx (hunk 3 above)
Comment (L-Sun) [reply packages/frontend/core/src/components/affine/setting-modal/general-setting/editor/general.tsx]: I have extracted above codes to another component. 

The following codes will refactored in another PR, where the `<SettingRow>` and `i18n` will be put in specific setting component. cc @JimmFly 
```ts
      <SettingRow
        name={t['com.affine.settings.editorSettings.general.ai.title']()}
        desc={t['com.affine.settings.editorSettings.general.ai.description']()}
      >
        <AISettings />
      </SettingRow>
      <SettingRow
        name={t['com.affine.appearanceSettings.font.title']()}
        desc={t['com.affine.appearanceSettings.font.description']()}
      >
        <FontFamilySettings />
      </SettingRow>
      <CustomFontFamilySettings />
      <SettingRow
        name={t[
          'com.affine.settings.editorSettings.general.default-new-doc.title'
        ]()}
        desc={t[
          'com.affine.settings.editorSettings.general.default-new-doc.description'
        ]()}
      >
        <NewDocDefaultModeSettings />
      </SettingRow>
```

File: packages/frontend/core/src/components/affine/setting-modal/general-setting/editor/general.tsx [hunk 9]
```diff
@@ -313,15 +314,49 @@ const NewDocDefaultModeSettings = () => {
...
+      // refresh the page to apply the change
+      window.location.reload();
+    },
+    [editorSettingService.editorSetting]
+  );
+  const onToggleAI = useCallback(
+    (checked: boolean) => {
+      openConfirmModal({
+        title: checked ? 'Enable AI?' : 'Disable AI?',
+        description: `Are you sure you want to ${checked ? 'enable' : 'disable'} AI? This will refresh the page to apply the change.`,
```
Comment (EYHN) [inline packages/frontend/core/src/components/affine/setting-modal/general-setting/editor/general.tsx]: i18n ?

Comment (JimmFly) [reply packages/frontend/core/src/components/affine/setting-modal/general-setting/editor/general.tsx]: The copy for this confirm modal is yet to be finalised. I will add the i18n once the copy is confirmed.

**Review by EYHN — APPROVED** [review]

File: packages/frontend/core/src/components/blocksuite/block-suite-editor/lit-adaper.tsx [hunk 10]
```diff
@@ -94,11 +99,15 @@ const usePatchSpecs = (page: Doc, shared: boolean, mode: DocMode) => {
     };
   }, [mode, page.collection, page.id]);
 
+  const [enableAI] = useState(
```
Comment (forehalo) [inline packages/frontend/core/src/components/blocksuite/block-suite-editor/lit-adaper.tsx]: is there any possibility that useState without set means nothing

Comment (EYHN) [reply packages/frontend/core/src/components/blocksuite/block-suite-editor/lit-adaper.tsx]: Just for read a constant value

equal with
```
const enableAI = useRef(xxx).current
```

Comment (forehalo) [reply packages/frontend/core/src/components/blocksuite/block-suite-editor/lit-adaper.tsx]: it's a one time reading, why useState, useRef, useXxx

File: packages/frontend/core/src/providers/modal-provider.tsx [hunk 11]
```diff
@@ -183,6 +184,44 @@ export const SignOutConfirmModal = () => {
...
+export const AIReloadConfirmModal = () => {
+  const editorSettingService = useService(EditorSettingService);
+  const enableAI = useLiveData(
+    editorSettingService.editorSetting.settings$
+  ).enableAI;
+  const [aiState, setAiState] = useState(enableAI);
+  const [open, setOpen] = useState(false);
+
+  useEffect(() => {
+    if (enableAI !== aiState) {
```
Comment (forehalo) [inline packages/frontend/core/src/providers/modal-provider.tsx]: `setOpen(enableAI !== aiState)`, if close and open it immediately, the confirm modal in other tabs could be closed automatically

File: packages/frontend/core/src/providers/modal-provider.tsx [hunk 12]
```diff
@@ -183,6 +184,44 @@ export const SignOutConfirmModal = () => {
...
+      setOpen(true);
+    }
+  }, [aiState, enableAI]);
+
+  const onConfirm = useCallback(() => {
+    setAiState(enableAI);
```
Comment (forehalo) [inline packages/frontend/core/src/providers/modal-provider.tsx]: page got reloaded, no need to set states

**Pull request merged**